### PR TITLE
Allow generic git repos over HTTPS

### LINKED ISSUE #72: Support other hosts than GitHub for https
The introduction of the "absolute imports" feature broke the ability to use a generic (non-github) git repo as a package URI.

Given that we have private Gitlab repos that house dependencies for our projects, this is a blocker. :-)

I'll have a PR up for it shortly.

### PR DESCRIPTION
**Pull request by cboggs**
This PR should resolve #72 and allow generic git-over-https dependencies. I've added some tests to ensure that various success paths are still working, as well as a couple of additional failure cases.

The Github-specific behaviors are still present in the form of the archive download optimizations, but there is no longer any up-front assumption that Github will be the only non-SSH dep source.

Would love your feedback! Thanks.

### CODE REVIEW

**Comment by sh0rez** [comment]
Hi, this is a super cool enhancement! Could you please rebase your changes onto current master? Merged a fix recently changed the same files as yours.

**Comment by cboggs** [comment]
Awesome! Merged master in and I _think_ I got the conflicts ironed out in a sane manner (tests are passing locally), but we'll see what the pipeline says!

File: spec/deps/git.go [hunk 1]
```diff
@@ -75,13 +75,13 @@ func (gs *Git) UnmarshalJSON(data []byte) error {
...
 
-// Name returns the repository in a go-like format (github.com/user/repo/subdir)
+// Name returns the repository in a go-like format (example.com/user/repo/subdir)
 func (gs *Git) Name() string {
 	return fmt.Sprintf("%s/%s/%s%s", gs.Host, gs.User, gs.Repo, gs.Subdir)
 }
 
 // LegacyName returns the last element of the packages path
-// example: github.com/ksonnet/ksonnet-lib/ksonnet.beta.4 becomes ksonnet.beta.4
+// example: example.com/ksonnet/ksonnet-lib/ksonnet.beta.4 becomes ksonnet.beta.4
```
Comment (sh0rez) [inline spec/deps/git.go]:
```suggestion
// example: github.com/ksonnet/ksonnet-lib/ksonnet.beta.4 becomes ksonnet.beta.4
```

This is actually located there, no need to change it

File: spec/deps/git.go [hunk 2]
```diff
@@ -100,14 +100,15 @@ func (gs *Git) Remote() string {
 
 // regular expressions for matching package uris
 const (
-	gitSSHExp     = `ssh://git@(?P<host>.+)/(?P<user>.+)/(?P<repo>.+).git`
-	gitSCPExp     = `^git@(?P<host>.+):(?P<user>.+)/(?P<repo>.+).git`
-	githubSlugExp = `github.com/(?P<user>[-_a-zA-Z0-9]+)/(?P<repo>[-_a-zA-Z0-9]+)`
+	gitSSHExp = `ssh://git@(?P<host>.+)/(?P<user>.+)/(?P<repo>.+).git`
+	gitSCPExp = `^git@(?P<host>.+):(?P<user>.+)/(?P<repo>.+).git`
+	// The long ugly pattern for ${host} here is a generic pattern for "valid URL with zero or more subdomains and a valid TLD"
+	gitSlugExp = `(?P<scheme>https://)?(?P<host>[a-zA-Z0-9][a-zA-Z0-9-\.]{1,61}[a-zA-Z0-9]\.[a-zA-Z]{2,})/(?P<user>[-_a-zA-Z0-9]+)/(?P<repo>[-_a-zA-Z0-9]+)`
```
Comment (sh0rez) [inline spec/deps/git.go]: `https://` is hardcoded. 

Either leave it as is, but remove the capture group (there is no point in capturing something that is always the same), or support `http` as well .... unsure if `http` git urls even exist

File: spec/deps/git.go [hunk 3]
```diff
@@ -100,14 +100,15 @@ func (gs *Git) Remote() string {
...
 )
 
 var (
-	gitSSHRegex     = regexp.MustCompile(gitSSHExp)
-	githubSlugRegex = regexp.MustCompile(githubSlugExp)
+	gitSSHRegex  = regexp.MustCompile(gitSSHExp)
```
Comment (sh0rez) [inline spec/deps/git.go]: Are these used at all? Feel like I forgot to remove them in my last PR

File: spec/deps/git.go (hunk 2 above)
Comment (sh0rez) [inline spec/deps/git.go]:
```suggestion
	gitHTTPSExp = `(?P<scheme>https://)?(?P<host>[a-zA-Z0-9][a-zA-Z0-9-\.]{1,61}[a-zA-Z0-9]\.[a-zA-Z]{2,})/(?P<user>[-_a-zA-Z0-9]+)/(?P<repo>[-_a-zA-Z0-9]+)`
```

This is basically a generic HTTP expression that matches full http:// urls and slugs. I think it is important to include the protocol

File: spec/deps/git.go [hunk 4]
```diff
@@ -167,6 +166,7 @@ func match(p string, exp string) (gs *Git, version string) {
 		gs.Host = matches["host"]
 		gs.User = matches["user"]
 		gs.Repo = matches["repo"]
+		gs.Scheme = matches["scheme"]
```
Comment (sh0rez) [inline spec/deps/git.go]: This is overridden anyways .. for SSH with `GitSchemeSSH` and for HTTPS with `GitSchemeHTTPS`.

I'd say just remove the whole scheme thing from the regex and handle it in the switch above

File: spec/deps/git.go (hunk 3 above)
Comment (cboggs) [reply spec/deps/git.go]: I might have goofed up the merge of master, the conflicts threw me for more of a loop than usual.  I'll check it out :-)

File: spec/deps/git.go (hunk 2 above)
Comment (cboggs) [reply spec/deps/git.go]: Agreed.

Comment (cboggs) [reply spec/deps/git.go]: :+1:

File: spec/deps/git.go (hunk 3 above)
Comment (cboggs) [reply spec/deps/git.go]: Looks like the SSH one was still in master and I missed it, which led me to leave mine there too. :-) Removed both!

File: spec/deps/git.go (hunk 4 above)
Comment (cboggs) [reply spec/deps/git.go]: :+1: Done deal.

**Comment by cboggs** [comment]
Ok, I think I got things cleaned up a bit and addressed the relevant areas. Thanks for the feedback!

**Comment by sc250024** [comment]
@cboggs Does this support the comment I mentioned here? https://github.com/jsonnet-bundler/jsonnet-bundler/issues/72#issuecomment-581470909

Basically, GitLab supports another level of `<org>/<group>/<repo>` instead of just the typical `<org>/<repo>` format.

File: .gitignore [hunk 5]
```diff
@@ -1 +1,2 @@
 _output/
+.vscode
```
Comment (metalmatze) [inline .gitignore]: You should use a global gitignore on your system rather then adding specifics to each project. Please remove from this PR.
Here's a guide on how to ignore things system wide: https://gist.github.com/subfuzion/db7f57fff2fb6998a16c

File: cmd/jb/main_test.go [hunk 6]
```diff
@@ -64,12 +64,12 @@ func TestParseDependency(t *testing.T) {
...
+			path: "git+ssh://git@example.com/jsonnet-bundler/jsonnet-bundler.git",
 			want: &deps.Dependency{
 				Source: deps.Source{
 					GitSource: &deps.Git{
 						Scheme: deps.GitSchemeSSH,
-						Host:   "github.com",
+						Host:   "example.com",
 						User:   "jsonnet-bundler",
 						Repo:   "jsonnet-bundler",
 						Subdir: "",
```
Comment (metalmatze) [inline cmd/jb/main_test.go]: I'd still like to keep one test with github.com, simply cause I've seen that 99% of imports are actually Github.

**Comment by sh0rez** [comment]
> @cboggs Does this support the comment I mentioned here? [#72 (comment)](https://github.com/jsonnet-bundler/jsonnet-bundler/issues/72#issuecomment-581470909)
> 
> Basically, GitLab supports another level of `<org>/<group>/<repo>` instead of just the typical `<org>/<repo>` format.

@sc250024 it does not properly. We are aware of this issue, but the fix is not trivial, because there is literally no way to differentiate between `host.tld/group/user/repo` and `host.tld/user/repo/subdir`. I think for your purpose it might still even work, although it parses the repo incorrectly (the `repo` bit ends up in subdir).

The most proper thing to use for now is the `ssh` style, because it has the `.git` path segment that splits repo and subdir

File: cmd/jb/main_test.go (hunk 6 above)
Comment (cboggs) [reply cmd/jb/main_test.go]: Oops, I thought I had left one in there - much agreed.

File: .gitignore (hunk 5 above)
Comment (cboggs) [reply .gitignore]: Oh yea, I spaced on that one. Thanks!

**Comment by sh0rez** [comment]
Wohoo :rocket:

**Pull request merged**